Mathlib/Combinatorics/Enumerative/Partition/GenFun.lean

```diff
@@ -0,0 +1,168 @@
+/-
+Copyright (c) 2025 Weiyi Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Weiyi Wang
+-/
+import Mathlib.Combinatorics.Enumerative.Partition.Basic
+import Mathlib.RingTheory.PowerSeries.PiTopology
+
+/-!
+# Generating functions for partitions
+
+This file defines generating functions related to partitions. Given a character function $f(i, c)$
+of a part $i$ and the number of occurrences of the part $c$, the related generating function is
+$$
+G_f(X) = \sum_{n = 0}^{\infty} \left(\sum_{p \in P_{n}} \prod_{i \in p} f(i, \#i)\right) X^n
+= \prod_{i = 1}^{\infty}\left(1 + \sum_{j = 1}^{\infty} f(i, j) X^{ij}\right)
+$$
+where $P_n$ is all partitions of $n$, $\#i$ is the count of $i$ in the partition $p$.
+We give the definition `Nat.Partition.genFun` using the first equation, and prove the second
+equation in `Nat.Partition.hasProd_genFun` (with shifted indices).
+
+This generating function can be specialized to
+* When $f(i, c) = 1$, this is the generating function for partition function $p(n)$.
+* When $f(i, 1) = 1$ and $f(i, c) = 0$ for $c > 1$, this is the generating function for
+    `#(Nat.Partition.distincts n)`.
+* When $f(i, c) = 1$ for odd $i$ and $f(i, c) = 0$ for even $i$, this is the generating function for
+    `#(Nat.Partition.odds n)`.
+(TODO: prove these)
+
+The definition of `Nat.Partition.genFun` ignores the value of $f(0, c)$ and $f(i, 0)$. The formula
+can be interpreted as assuming $f(i, 0) = 1$ and $f(0, c) = 0$ for $c \ne 0$. In theory we could
+respect the actual value of $f(0, c)$ and $f(i, 0)$, but it makes the otherwise finite sum and
+product potentially infinite.
+-/
+
+open Finset PowerSeries
+open scoped PowerSeries.WithPiTopology
+
+namespace Nat.Partition
+variable {R : Type*} [CommSemiring R]
+
+/-- Generating function associated with character $f(i, c)$ for partition functions, where $i$ is a
+part of the partition, and $c$ is the count of that part in the partition. The character function is
+multiplied within one `n.Partition`, and summed among all `n.Partition` for a fixed `n`. This way,
+each `n` is assigned a value, which we use as the coefficients of the power series. -/
+def genFun (f : ℕ → ℕ → R) : R⟦X⟧ :=
+  PowerSeries.mk fun n ↦ ∑ p : n.Partition, ∏ i ∈ p.parts.toFinset, f i (p.parts.count i)
```

Comment:
I wonder if it's easier to express this using a Finsupp.prod, combined with Multiset.toFinsupp?

Reply:
The prod part turns out defeq to `p.parts.toFinsupp.prod fun i c ↦ f i c`. Let me change to it after checking there is no weird defeq in proofs

Reply:
That should be ` p.parts.toFinsupp.prod f`, unless there's some subtyping I'm missing